### ISSUE TITLE
jupyter connection to kernel hanging forever

### ISSUE THREAD

**Issue by arita37** [jupyter/notebook#5601]
Currently running in docker Jupyter:

```
1251 root      20   0  690572  67080  15280 S   0.0   0.2   0:08.54 /usr/local/bin/python /usr/local/bin/jupyter-lab --port=8888 --no-browser --ip=0.0.0.0 --debug --no-browser -+
 2551 root      20   0  539072  46808  13568 S   0.0   0.1   0:00.36 /usr/local/bin/python -m ipykernel_launcher -f /root/.local/share/jupyter/runtime/kernel-8bc2b516-4e8a-4f86-8+
 
```

Jupyter can launch sucessfully yhe ipykernel, if we ccheck manually, it works :

```
#### Testing the kernel
from jupyter_client import BlockingKernelClient ; client = BlockingKernelClient()
client.load_connection_file('/root/.local/share/jupyter/runtime/kernel-8bc2b516-4e8a-4f86-8a06-546b57051846.json')
def _output_hook(msg):  
  print(msg) ;
client.start_channels()  ; client.execute_interactive("import os; print(os)", output_hook=_output_hook,)  

```

BUT, in jupyter notebook, connection to kernel failed and hanging forevever 

Suppose this is related to Tornado or some security issues.

```
4.json']
[D 14:36:17.510 LabApp] Connecting to: tcp://127.0.0.1:36699
[I 14:36:17.511 LabApp] Kernel restarted: be3a274a-dca2-4d10-b904-12660b8e6c24
[D 14:36:17.511 LabApp] Connecting to: tcp://127.0.0.1:49390
[D 14:36:17.512 LabApp] activity on be3a274a-dca2-4d10-b904-12660b8e6c24: status (busy)
[D 14:36:17.513 LabApp] activity on be3a274a-dca2-4d10-b904-12660b8e6c24: shutdown_reply
[D 14:36:17.513 LabApp] activity on be3a274a-dca2-4d10-b904-12660b8e6c24: status (idle)
[D 14:36:17.875 LabApp] Kernel info reply received: be3a274a-dca2-4d10-b904-12660b8e6c24
[D 14:36:17.875 LabApp] 200 POST /api/kernels/be3a274a-dca2-4d10-b904-12660b8e6c24/restart?1594478177309 (100.66.227.2) 569.17ms
[D 14:36:21.104 LabApp] Accepting token-authenticated connection from 100.66.227.2
[D 14:36:21.106 LabApp] 200 GET /api/contents/?content=1&1594478181103 (100.66.227.2) 2.13ms



```

Install is as follow 

```

Installed software:
Python 3.8.2
pip 20.0.2 from /usr/local/lib/python3.8/site-packages/pip (python 3.8)
jupyter core     : 4.6.3
jupyter-notebook : 6.0.3
qtconsole        : not installed
ipython          : 7.13.0
ipykernel        : 5.2.1
jupyter client   : 6.1.3
jupyter lab      : 1.2.6
nbconvert        : 5.6.1
ipywidgets       : 7.5.1
nbformat         : 5.0.6
traitlets        : 4.3.3
Node v12.16.2
NPM 6.14.4

Installed Python packages:
Package            Version     
------------------ ------------
attrs              19.3.0      
backcall           0.1.0       
bleach             3.1.4       
bokeh              2.0.2       
certifi            2020.4.5.1  
chardet            3.0.4       
colorama           0.4.3       
cycler             0.10.0      
decorator          4.4.2       
defusedxml         0.6.0       
entrypoints        0.3         
gitdb              4.0.4       
GitPython          3.1.1       
idna               2.9         
imageio            2.8.0       
ipykernel          5.2.1       
ipython            7.13.0      
ipython-genutils   0.2.0       
ipywidgets         7.5.1       
jedi               0.17.0      
Jinja2             2.11.2      
joblib             0.14.1      
json5              0.9.4       
jsonschema         3.2.0       
jupyter-client     6.1.3       
jupyter-core       4.6.3       
jupyterlab         1.2.6       
jupyterlab-git     0.10.0      
jupyterlab-latex   2.0.0       
jupyterlab-server  1.0.7       
kiwisolver         1.2.0       
MarkupSafe         1.1.1       
matplotlib         3.2.1       
mistune            0.8.4       
mpmath             1.1.0       
nbconvert          5.6.1       
nbdime             1.1.0       
nbformat           5.0.6       
networkx           2.4         
nose               1.3.7       
notebook           6.0.3       
numexpr            2.7.1       
numpy              1.18.3      
packaging          20.3        
pandas             1.0.3       
pandocfilters      1.4.2       
parso              0.7.0       
patsy              0.5.1       
pexpect            4.8.0       
pickleshare        0.7.5       
Pillow             7.1.1       
pip                20.0.2      
plotly             4.6.0       
prometheus-client  0.7.1       
prompt-toolkit     3.0.5       
ptyprocess         0.6.0       
Pygments           2.6.1       
pyparsing          2.4.7       
pyrsistent         0.16.0      
python-dateutil    2.8.1       
pytz               2019.3      
PyWavelets         1.1.1       
PyYAML             5.3.1       
pyzmq              19.0.0      
requests           2.23.0      
retrying           1.3.3       
scikit-image       0.16.2      
scikit-learn       0.22.2.post1
scipy              1.4.1       
seaborn            0.10.0      
Send2Trash         1.5.0       
setuptools         46.1.3      
six                1.14.0      
smmap              3.0.2       
sympy              1.5.1       
terminado          0.8.3       
testpath           0.4.4       
tornado            6.0.4       
traitlets          4.3.3       
typing-extensions  3.7.4.2     
urllib3            1.25.9      
wcwidth            0.1.9       
webencodings       0.5.1       
wheel              0.34.2      
widgetsnbextension 3.5.1       



```

**Comment by kevin-bates** [comment]
Hi @arita37 - being that the kernel just successfully restarted the kernel (for whatever reason), have you tried forcing a reconnect to the kernel.  In Lab this is a bit more involved, so I've provided a screenshot.
![image](https://user-images.githubusercontent.com/22599560/87226760-cd71b980-c34a-11ea-9ba7-65de67dcbba6.png)

If the restart caused new ports to be used, there's a known issue in Lab that it doesn't reconnect to the WebSocket following restarts. (See https://github.com/jupyterlab/jupyterlab/pull/8432)

Prior to the restart, was the kernel usable, or do you see it has "hanging forever" from the start?  If usable and you didn't intentionally restart the kernel, then you might want to try to determine why it restarted.  Perhaps there's a memory consumption issue (that might appear as a hung process) or something to that effect, the system needed to terminate that process - at which time the notebook server performs its auto-restart sequence.

I'm not sure how a security issue is being manifested in this case.

**Comment by arita37** [comment]
Obviously, kernel was re-started 100s of times...
And lookong for solution for days....


I can connect manually to the kernel 
and run python code.


But , jupyter notebook cannot.

Reasons ?

1) Tornado ?
2) security issue (am on docker in linux server)
3) another bug

**Comment by kevin-bates** [comment]
>Obviously, kernel was re-started 100s of times...

I'm sorry, but It's not at all _obvious_ that the kernel has restarted "100s of times" from anything you shared previously.  
> I can connect manually to the kernel

I'm surprised the kernel is staying up long enough for you to connect, but since its restarting with the same ports, I suppose that could work.  From where are you running the BlockingKernelClient code - inside or outside the container?

Do you know if the restarts are triggered by the notebook sending code or are they happening on their own, w/o doing anything within the notebook?

Please provide your complete docker log after at least a couple of restarts have occurred.  If it happens to be 100+ restarts, fine, include them all.

**Comment by arita37** [comment]
The python is run inside the container.
and can see the print.

1) there is no automatic restart :
   Just manul ones

2) Jupyter just says:
  “Connecting”  forever

Suspect this is security issue on the linux server
Which hosting the docker.

**Comment by arita37** [comment]
Used Jupyter docker is :
   https://github.com/amalic/Jupyterlab

This docker jupyter works in no secure environnment.

**Comment by kevin-bates** [comment]
Thanks for the additional information.  I'm not seeing an issue with the docker image.  I can start and use a python kernel just fine.  You might be best off opening an issue in the referenced repository since perhaps there's a relation to the host on which the container is running.  My host is macOS 10.15.5.

Until there's a concrete error or problem displayed, there really isn't much help we can provide.

**Comment by arita37** [comment]
As explained, am on internal server with security, There is some unknown security firewall preventing :

 Jupyter notebook server to communicate
with ipykernel. Error message is
    “Kernel Connecting” and the notebook is hanging.

   2) If i try manually to connect with python code above, to ipykernel , it works.


Question :
   1) Where the communication process
       between jupyter notebook and ipykernel is described or the source code ?

    2) Does it involve tornado ?

    3) Any security protection which prevent notebook to execute code ?



Thanks

**Comment by arita37** [comment]
@kevin-bates 

Is this where the connection code is located ?
https://github.com/jupyter/jupyter_server/blob/ae0fde37215d9e11f80bb602c9819c7e7a7cd81a/tests/services/kernels/test_api.py

**Comment by kevin-bates** [comment]
No.  That's test code in `jupyter_server` which isn't really used yet.  I think Jupyter Lab is looking to start using `jupyter_server`, rather than notebook, in its 3.0 release.  `jupyter_server` is essentially a copy of the notebook repo sans front-end code - it's purely the server.

For the server-side of things, you'll probably want to focus on the [kernel handlers](https://github.com/jupyter/notebook/blob/master/notebook/services/kernels/handlers.py) within notebook, and the [ZMQChannelsHandler](https://github.com/jupyter/notebook/blob/master/notebook/services/kernels/handlers.py#L93) in particular.  

I'm not very familiar with the front-end, but I believe you'd find the kernel API usage in the [kernels services "package"](https://github.com/jupyter/notebook/tree/master/notebook/static/services/kernels).

Finally, your journey will very likely lead you to [`jupyter_client`](https://github.com/jupyter/jupyter_client/tree/master/jupyter_client) as well.

**Comment by arita37** [comment]
Thanks very much for the pointers.


My guess is this might be related to authentication or whatever security check:
(Since manual kernel connection works...)

class KernelActionHandler(APIHandler):

    @web.authenticated

**Comment by arita37** [comment]
Finally , this is the same problem as decribed here:
https://github.com/jupyter/notebook/issues/2664

If I create ssh tunnel to the jupyter server port :  
     Can launch kernel
otherwise kernel hanging forever.....

This kind of websocker machanism should be more documented in jupyter kernel.... side

**Comment by kevin-bates** [comment]
Awesome - I'm glad to hear you're moving forward.
> This kind of websocker machanism should be more documented in jupyter kernel.... side

Contributions are always welcome!  Seems fitting for a troubleshooting or configuration-related item.

I'm closing this issue as a duplicate of #2664.

**Comment by arita37** [comment]
There is a mechanism on websocker side
Which is blocked by company server firewall.

If we do a reverse tunnel SSL,
We can communicate with kernel.
But, not clear  at which part of the code
the blocking happens
